Modules/modFileEncoding.ACM

```diff
@@ -0,0 +1,257 @@
+  
+  Dim c As Long
+  
+  UTF8CP_to_UCS2CP = 0 ' intitally set to zero (fail)
+  
+  If P > UBound(bytsUTF8) Then Exit Function
+  If P < LBound(bytsUTF8) Then Exit Function
+  
+  ReDim bytsUCS2(1) ' always 2 bytes
+  
+  If (bytsUTF8(P) And 128) = 0 Then '128 = 1000 0000 'this byte is 0xxx xxxx
+    ' We have a single byte UTF8
+    bytsUCS2(0) = bytsUTF8(P)
+    bytsUCS2(1) = 0
+    UTF8CP_to_UCS2CP = 1
+  ElseIf (bytsUTF8(P) And 224) = 192 Then  '224 = 1110 0000, 192 = 1100 0000 'this byte is 110x xxxx
+    If UBound(bytsUTF8) < P + 1 Then Exit Function 'check we have enough bytes left
+    If (bytsUTF8(P + 1) And 192) <> 128 Then Exit Function '192 = 1100 0000, 128 = 1000 0000 2nd byte must be 10xx xxxx
+    ' We have a valid 2 byte UTF8
+    c = bit16ShiftR((bytsUTF8(P) And 31), -6) '31 = 0001 1111
+    c = c Or (bytsUTF8(P + 1) And 63) '63 = 0011 1111
+    bytsUCS2(0) = c And 255 '255 = 1111 1111  least significant byte
+    bytsUCS2(1) = bit16ShiftR(c, 8) ' most significant Byte
+    UTF8CP_to_UCS2CP = 2
+  ElseIf (bytsUTF8(P) And 240) = 224 Then '240 = 1111 0000, 224 = 1110 0000 'this byte is 1110 xxxx
+    If UBound(bytsUTF8) < P + 2 Then Exit Function 'check we have enough bytes left
+    If (bytsUTF8(P + 1) And 192) <> 128 Then Exit Function '2nd byte must be 10xx xxxx
+    If (bytsUTF8(P + 2) And 192) <> 128 Then Exit Function '3rd byte must be 10xx xxxx
+    ' We have a valid 3 byte UTF8
+    c = bit16ShiftR((bytsUTF8(P) And 15), -12) '15 = 0000 1111
+    c = c Or bit16ShiftR((bytsUTF8(P + 1) And 63), -6) '63 = 0011 1111
+    c = c Or (bytsUTF8(P + 2) And 63) '63 = 0011 1111
+    bytsUCS2(0) = c And 255 '255 = 1111 1111  least significant byte
+    bytsUCS2(1) = bit16ShiftR(c, 8) ' most significant Byte
+    UTF8CP_to_UCS2CP = 3
+  Else
+    ' we cannot deal with 4 byte UTF8 so we get an error
+    Exit Function
+  End If
+    
+End Function
+
+Private Function UCS2CP_to_UTF8CP(bytsUCS2() As Byte, bytsUTF8() As Byte) As Boolean
+  ' This function takes a 2 byte zero based array encoded in UCS2 and returns a 1 to 3 byte array encoded in UTF8
+  
+  Dim i As Long ' this is used to represent the character we have
+  Dim j As Long
+  
+  UCS2CP_to_UTF8CP = False ' intially set return value to be false
+  
+  If LBound(bytsUCS2) <> 0 Then Exit Function
+  If UBound(bytsUCS2) <> 1 Then Exit Function
+  
+  i = bytsUCS2(1) * 256 + bytsUCS2(0) ' Little Endian
+  
+  If i < 128 Then '128 = &H80
+    ' encoded in 1 byte in UTF8
+    ReDim bytsUTF8(0)
+    bytsUTF8(0) = i
+    UCS2CP_to_UTF8CP = True
+  ElseIf i < 2048 Then '2048 = &H0800
+    ' encoded in 2 bytes in UTF8
+    ReDim bytsUTF8(1)
+    bytsUTF8(1) = (i And 63) Or 128 ' 63 = 00111111, 128 = 10000000
+    i = bit16ShiftR(i, 6)
+    bytsUTF8(0) = (i And 31) Or 192 ' 31 = 00011111, 192 = 11000000
+    UCS2CP_to_UTF8CP = True
+  Else
+    ' encoded in 3 bytes in UTF8
+    ReDim bytsUTF8(2)
+    bytsUTF8(2) = (i And 63) Or 128 ' 63 = 0011 1111, 128 = 1000 0000

```

Comment:
Mix of binary and hex in comments?

Reply:
This is deliberate. I have added some explanation of UTF8 formatting to assist with understanding the logic.